File: 2/2.py
import sys
import os
from typing import List

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

from utils.operate_on_lines import operate_on_lines


def same_sign(number1, number2):
    return number1 * number2 >= 0


def valid_range(diff):
    return abs(diff) <= 3 and abs(diff) >= 1


def get_numbers(line: str):
    return [int(i) for i in line.split(" ")]


def is_safe(numbers):
    first_diff = numbers[1] - numbers[0]

    for i in range(len(numbers) - 1):
        diff = numbers[i + 1] - numbers[i]

        if not same_sign(first_diff, diff):
            return False

        if not valid_range(diff):
            return False

    return True


def two_a():
    total_safe = [0]

    def fn(stripped_line: str):
        numbers = get_numbers(stripped_line)

        if is_safe(numbers):
            total_safe[0] += 1

    operate_on_lines(fn)

    print(total_safe[0])


def identify_dominant_number_sign(numbers: List[int]):
    if len(numbers) < 4:
        return None
    start_diffs = [numbers[i + 1] - numbers[i] for i in range(3)]
    start_diffs.sort()
    return start_diffs[1]


def two_b():
    total_safe = [0]

    def fn(stripped_line: str):
        numbers = get_numbers(stripped_line)

        if is_safe(numbers):
            total_safe[0] += 1
            return

        for i in range(len(numbers)):
            new_numbers = numbers[:i] + numbers[i + 1 :]
            safe = is_safe(new_numbers)
            if safe:
                total_safe[0] += 1
                return

    operate_on_lines(fn)

    print(total_safe[0])


if __name__ == "__main__":
    two_b()

File: 2/2.cpp
#include "2.h"

#include <iostream>
#include <memory>
#include <vector>

#include "filing.h"

std::vector<int> get_numbers(std::string line) {
  std::vector<int> numbers;
  std::string_view view(line);

  size_t start = 0;
  for (size_t i = 0; i < view.length(); ++i) {
    char current_char = view[i];

    if (current_char == ' ') {
      numbers.push_back(std::stoi(std::string(view.substr(start, i))));
      start = i + 1;
    }
  }

  numbers.push_back(std::stoi(std::string(view.substr(start, view.length()))));

  return std::move(numbers);
}

bool same_sign(int n1, int n2) { return n1 * n2 >= 0; }

bool valid_range(int diff) { return abs(diff) <= 3 && abs(diff) >= 1; }

bool is_safe(std::vector<int> numbers) {
  int first_diff = numbers[1] - numbers[0];

  for (size_t i = 0; i < numbers.size() - 1; ++i) {
    int diff = numbers[i + 1] - numbers[i];

    if (!same_sign(first_diff, diff)) {
      return false;
    }

    if (!valid_range(diff)) {
      return false;
    }
  }

  return true;
}

void two_a() {
  std::ifstream file = getMainFile(2);

  int safe_lines = 0;
  std::string line;
  while (std::getline(file, line)) {
    std::vector<int> numbers = get_numbers(line);

    if (is_safe(numbers)) {
      safe_lines += 1;
    }
  }

  std::cout << safe_lines << "\n";
}

void two_b() {
  std::ifstream file = getMainFile(2);

  int safe_lines = 0;
  std::string line;
  while (std::getline(file, line)) {
    std::vector<int> numbers = get_numbers(line);

    if (is_safe(numbers)) {
      safe_lines += 1;
      continue;
    }

    for (size_t i = 0; i < numbers.size(); i++) {
      std::vector<int> sectioned_numbers;

      for (size_t j = 0; j < i; ++j) {
        sectioned_numbers.push_back(numbers[j]);
      }

      for (size_t j = i + 1; j < numbers.size(); ++j) {
        sectioned_numbers.push_back(numbers[j]);
      }

      if (is_safe(sectioned_numbers)) {
        safe_lines += 1;
        break;
      }
    }
  }

  std::cout << safe_lines << "\n";
}

void two() { two_b(); }